src/snowflake/cli/api/utils/definition_rendering.py

```diff
@@ -0,0 +1,251 @@
+                current_dict_level[var] = self.rendered_value
+            else:
+                current_dict_level.setdefault(var, {})
+                current_dict_level = current_dict_level[var]
+
+    def read_from_context(self, context: dict[str, Any]) -> Any:
+        """
+        Takes a multi-level context dict as input.
+
+        If the variable has multi-levels (e.g. ctx.env), recursively traverse the dictionary
+        to find the key that the variable points to.
+
+        Returns the value in that location.
+
+        Raise UndefinedError if the variable is None or not found.
+        """
+        current_dict_level = context
+        for key in self._vars_chain:
+            if (
+                not isinstance(current_dict_level, dict)
+                or key not in current_dict_level
+            ):
+                raise UndefinedError(f"Could not find template variable {self.key}")
+            current_dict_level = current_dict_level[key]
+
+        value = current_dict_level
+        if value is None or isinstance(value, (dict, list)):
+            raise UndefinedError(
+                f"Template variable {self.key} does not contain a valid value"
+            )
+
+        return value
+
+    def __hash__(self):
+        return hash(self.key)
+
+    def __eq__(self, other):
+        return self.key == other.key
+
+
+def _get_referenced_vars(
+    ast_node: nodes.Template,
+    template_value: str,
+    current_attr_chain: Optional[list[str]] = None,
+) -> set[TemplateVar]:
+    """
+    Traverse Jinja AST to find the variable chain referenced by the template.
+    A variable like ctx.env.test is internally represented in the AST tree as
+    Getattr Node (attr='test') -> Getattr Node (attr='env') -> Name Node (name='ctx')
+    """
+    all_referenced_vars: set[TemplateVar] = set()
+    if isinstance(ast_node, nodes.Getattr):
+        current_attr_chain = [ast_node.attr] + (current_attr_chain or [])  # type: ignore[attr-defined]
+    elif isinstance(ast_node, nodes.Name):
+        current_attr_chain = [ast_node.name] + (current_attr_chain or [])  # type: ignore[attr-defined]
+        all_referenced_vars.add(TemplateVar(current_attr_chain))
+        current_attr_chain = None
+    elif (
+        not isinstance(ast_node, (nodes.Template, nodes.TemplateData, nodes.Output))
+        or current_attr_chain is not None
+    ):
+        raise InvalidTemplate(f"Unexpected templating syntax in {template_value}")
+
+    for child_node in ast_node.iter_child_nodes():
+        all_referenced_vars.update(
+            _get_referenced_vars(child_node, template_value, current_attr_chain)
+        )
+
+    return all_referenced_vars
+
+
+def _get_referenced_vars_from_str(
+    env: Environment, template_str: str
+) -> set[TemplateVar]:
+    ast = env.parse(template_str)
+    return _get_referenced_vars(ast, template_str)
+
+
+def _build_dependency_graph(
+    env: Environment, all_vars: set[TemplateVar], context: dict[str, Any]
```

Comment:
```suggestion
    env: Environment, all_vars: set[TemplateVar], context: Context
```
Where `Context = dict[str, Any]`